src/engine/renderer/tr_image.cpp

```diff
@@ -833,13 +896,61 @@ void R_UploadImage( const byte **dataArray, int numLayers, int numMips,
 	scaledWidth = image->width;
 	scaledHeight = image->height;
 
+	int picMin = r_picMin->integer;
```

Comment:
The logic to decide mip level is getting very long. Can you put it in a separate function?

Reply:
Done